Updates documentation

rickstaa opened the pull request:
This pull request is a response on https://github.com/reviewdog/action-remark-lint/pull/25. In this pull request, I added a format feature to the repository. After a discussion, @haya14busa and I decided that this feature should be in a seperate repository. I, therefore, removed the formatting related changes and kept the rest of the changes. This results in the following features.

:sparkles: Main changes
- Better gh-action tests.
- Updates action documentation.

:boom: Breaking changes

- I changed the dash shell to the bash shell into the docker container. This allows us to both catches the errors and have cleaner code. This doesn't change the docker size by much.
- I changed the `black_flags` input argument to `black_args` as we can also pass non-flag arguments like file paths.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
+remark_exit_val="0"
+reviewdog_exit_val="0"
+echo "[action-remark-lint] Checking markdown code using the remark-lint linter and reviewdog..."
+remark_lint_output="$(remark --frail --quiet --use=remark-preset-lint-recommended ${remark_args[*]} 2>&1)" ||
```
I can not do this without globbing or word splitting.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
+    -fail-on-error="${INPUT_FAIL_ON_ERROR,,}" \
+    -level="${INPUT_LEVEL}" \
+    -tee \
+    ${INPUT_REVIEWDOG_FLAGS} || reviewdog_exit_val="$?"
```
I can not do this without globbing and word splitting.

haya14busa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
+    -name="${INPUT_TOOL_NAME}" \
+    -reporter="${INPUT_REPORTER}" \
+    -filter-mode="${INPUT_FILTER_MODE}" \
+    -fail-on-error="${INPUT_FAIL_ON_ERROR,,}" \
```
What does `,,` do? typo?

rickstaa replied:
@haya14busa It's bash 4.0 syntax for making strings lowercase (see this StackOverflow post). I will add a comment such that it doesn't confuse other developers.

haya14busa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
+fi
+
+# Throw error if an error occurred and fail_on_error is true
+if [[ "${INPUT_FAIL_ON_ERROR,,}" = 'true' && ("${remark_error}" = 'true' || \
```
ditto

haya14busa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
-set -e # Exit immediately if a command exits with a non-zero status
+#!/bin/bash
+set -eu # Increase bash strictness
+set -o pipefail
```
Do we need this pipefail?
If we unset pipefail, you can use `PIPESTATUS` bash feature to get each command in pipe instead of capturing `remark_lint_output`.

limakzi replied:
@haya14busa I would use pipefail with proper pipefailure.

rickstaa replied:
@haya14busa I tried using PIPESTATUS previously, but I run into problem when the Reviewdog help dialogue is returned due to incorrect (user supplied) Reviewdog flags (see my [stackoverflow issue](https://stackoverflow.com/questions/65485359/redirecting-stderr-changes-exit-codes-of-piped-command)). This was however related to me wanting to access unchanged exit code of black while it was piped into Reviewdog. As I now first store the black output in a seperate variable the pipefail can indeed be replaced with using PIPESTATUS. @limakzi Why do you suggest using pipefail over PIPESTATUS?

rickstaa replied:
@haya14busa and @limakzi I removed the error catching behavoir after all since it is not needed if we don't do the formatting.

haya14busa commented on .github/workflows/test.yml:
```diff
@@ -15,6 +15,16 @@ jobs:
           reporter: github-check
           level: info
+          workdir: ./testdata/
+      - name: Check if code is left untouched (not formatted)
```
I think we can remove this test if the action doesn't intend to change the code in the first place.
I don't think we need to copy&past this test for each reporter at least.

rickstaa replied:
@haya14busa Your right the remark linter doesn't change the files unless the user explicitly adds the `--output` flag. This is therfore default behavoir which we don't need to check.

limakzi commented on README.md:
```diff
@@ -31,80 +32,53 @@ jobs:
 
-This action updates major/minor release tags on a tag push. e.g. Update v1 and v1.2 tag when released v1.2.3.
-ref: <https://help.github.com/en/articles/about-actions#versioning-your-action>
+**Optional**. Report level for reviewdog \[info, warning, error]. It's same as `-level` flag of reviewdog. Defaults to `error`.
```
```suggestion
**Optional**. Report level for reviewdog `\[info, warning, error]`. It's same as `-level` flag of reviewdog. Defaults to `error`.
```

rickstaa replied:
@limakzi Good catch! Thanks a lot for reading through the docs to find these syntax errors!

limakzi commented on README.md:
```diff
@@ -31,80 +32,53 @@ jobs:
 
-This reviewdog action template itself is integrated with reviewdog to run lints
-which is useful for Docker container based actions.
+**Optional**. Reporter of reviewdog command \[github-pr-check, github-pr-review, github-check].
```
```suggestion
**Optional**. Reporter of reviewdog command `\[github-pr-check, github-pr-review, github-check]`.
```

limakzi commented on README.md:
```diff
@@ -31,80 +32,53 @@ jobs:
-This reviewdog action template itself is integrated with reviewdog to run lints
-which is useful for Docker container based actions.
+**Optional**. Reporter of reviewdog command \[github-pr-check, github-pr-review, github-check].
+Default is github-pr-check. github-pr-review can use Markdown and add a link to rule page in reviewdog reports.
```
```suggestion
Default is `github-pr-check`. `github-pr-review` can use Markdown and add a link to rule page in reviewdog reports.
```

limakzi commented on README.md:
```diff
@@ -31,80 +32,53 @@ jobs:
+### `filter_mode`
 
-Supported linters:
+**Optional**. Filtering mode for the reviewdog command \[added, diff_context, file, nofilter]. Defaults to `"added"`.
```
```suggestion
**Optional**. Filtering mode for the reviewdog command `\[added, diff_context, file, nofilter]`. Defaults to `added`.
```

limakzi commented on action.yml:
```diff
@@ -2,30 +2,36 @@ name: "Run remark-lint with reviewdog"
     description: |
-      Reporter of reviewdog command [github-pr-check,github-pr-review,github-check].
-      Default is github-pr-check.
-      github-pr-review can use Markdown and add a link to rule page in reviewdog reports.
+      Reporter of reviewdog command [github-pr-check, github-pr-review, github-check].
+      Defaults to 'github-pr-check'.
```
I would keep the same format of description everywhere.
So, either we write in two lines.

```
description: |
<Desc>
<Default>
```

or
```
description:  <Desc> <Default>
```

rickstaa replied:
@limakz I think your right. Neverteless do you think we really need to explicitly state the default inside the description tag. The default is already stated in both the README.md and can be found inside the `defaults` tag.

rickstaa replied:
@limakzi I went ahead and made these changes see cdae7e2. I can revert if you think adding it is still benificial.

limakzi commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
+fi
+
+# Throw error if an error occurred and fail_on_error is true
+if [[ "${INPUT_FAIL_ON_ERROR,,}" = 'true' && ("${remark_error}" = 'true' || \
```
`remark_exit_val` has the exact same meaning as `remark_error`.

rickstaa replied:
Your right! Creating remark_error is redundant is now redundant in this version. Thanks!

rickstaa replied:
@limakzi @haya14busa Great I think I can now remove many lines as both remark and reviewdog only throw 0 and 1 exit codes. I previously also checked for other exit codes but they should not occur. I therefore think all these checks can be removed.

rickstaa commented on action.yml:
```diff
@@ -2,30 +2,36 @@ name: "Run remark-lint with reviewdog"
+    default: ${{ github.token }}
   tool_name:
-    description: "Tool name to use for reviewdog reporter"
+    description: "Tool name to use for reviewdog reporter. Defaults to 'black-format'."
```
Good catch! :)

rickstaa commented on action.yml:
```diff
@@ -2,30 +2,36 @@ name: "Run remark-lint with reviewdog"
     default: ""
+  # Reviewdog related inputs
+  github_token:
+    description: "The automatically created secret github action token."
```
Good point I changed it to `description: "GITHUB_TOKEN. Defaults to '${{ github.token }}'."` to be more in line with the other Reviewdog actions.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,77 @@
 export REVIEWDOG_GITHUB_API_TOKEN="${INPUT_GITHUB_TOKEN}"
 
-remark --frail --quiet --use=remark-preset-lint-recommended . ${INPUT_REMARK_FLAGS} 2>&1 |
+# If no arguments are given use current working directory
```
@haya14busa I think they both can be removed. I think they were added by @prologic, @asbjornu to make sure that the errors were passed through to Reviewdog. This bug was, however fixed by adding the right parsing syntax to the [reviewdog/errorformat repository(https://github.com/reviewdog/errorformat/commit/b7afc78df3963ef237a0f51a3ca19b5130ff94b6).

rickstaa commented:
@haya14busa, @limakzi Thanks for all the suggestions. Based on your comments, I realized that the error catching behavoir is not needed anymore if we are not adding the formatting feature. I therefore trimmed down the code and it is now very close to the original code on the master branch. The features that are now included are:

:sparkles: Features
- The documentation was updated.
- The action now also works when people incorrectly capitalize input arguments (e.g. `TRUE` instead of `true`).
- The action now also works with `args`.

:boom: Breaking changes
- I changed the dash shell to the bash shell into the docker container. This allows for cleaner code. The docker size doesn't change much.

I think both features are not really required, but I added them when creating the auto-format feature. We can therefore also close this pull-request and keep the original version (dash shell, without dev args and strict input arg checking). Let me know what you think.

rickstaa commented:
🧪 Additional (succesfull) test results can be found https://github.com/rickstaa/action-test-repo/pull/52.

haya14busa commented on README.md:
```diff
@@ -28,83 +29,56 @@ jobs:
+**Optional**. The directory to run remark-lint in. Defaults to `.`.
 
-### remark_flags
+### `format`
```
remove?

rickstaa replied:
Done.

haya14busa commented on entrypoint.sh:
```diff
@@ -1,19 +1,30 @@
-  ${INPUT_REVIEWDOG_FLAGS}
\ No newline at end of file
+    -name="${INPUT_TOOL_NAME}" \
+    -reporter="${INPUT_REPORTER,,}" \
```
Ditto for other flags.

rickstaa replied:
Good point. I removed the lowercase syntax all together as even with fail-on-error this might not work if we change the fail-on-error behaviour in the future.

rickstaa commented on README.md:
```diff
@@ -28,83 +29,56 @@ jobs:
-This action updates major/minor release tags on a tag push. e.g. Update v1 and v1.2 tag when released v1.2.3.
-ref: <https://help.github.com/en/articles/about-actions#versioning-your-action>
+**Optional**. Reporter of reviewdog command `[github-pr-check, github-pr-review, github-check]`.
+Default is `github-pr-check`. `github-pr-review` can use Markdown and add a link to rule page in reviewdog reports.
```
Your right this is still from one of the first versions of the documentation.

rickstaa replied:
I removed the full statement as people should already be familiar with the reviewdog flags or look them up in the documentation.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,21 @@
-remark --frail --quiet --use=remark-preset-lint-recommended . ${INPUT_REMARK_FLAGS} 2>&1 |
+# NOTE: ${VAR,,} Is bash 4.0 syntax to make strings lowercase.
+echo "[action-remark-lint] Checking markdown code with the remark-lint linter and reviewdog..."
+remark --use=remark-preset-lint-recommended . ${remark_args} 2>&1 |
```
Done by design.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,21 @@
-remark --frail --quiet --use=remark-preset-lint-recommended . ${INPUT_REMARK_FLAGS} 2>&1 |
+# NOTE: ${VAR,,} Is bash 4.0 syntax to make strings lowercase.
+echo "[action-remark-lint] Checking markdown code with the remark-lint linter and reviewdog..."
+remark --use=remark-preset-lint-recommended . ${remark_args} 2>&1 |
```
Good catch!

rickstaa commented on Dockerfile:
```diff
@@ -2,7 +2,9 @@ FROM prologic/remark-lint:latest
 ENV REVIEWDOG_VERSION=v0.11.0-nightly20201213+85edbc6
 
-RUN wget -O - -q https://raw.githubusercontent.com/reviewdog/nightly/master/install.sh| sh -s -- -b /usr/local/bin/ ${REVIEWDOG_VERSION}
+RUN apk add --no-cache bash
```
This will be fixed in next commit.

rickstaa commented on Dockerfile:
```diff
@@ -2,7 +2,9 @@ FROM prologic/remark-lint:latest
-RUN wget -O - -q https://raw.githubusercontent.com/reviewdog/nightly/master/install.sh| sh -s -- -b /usr/local/bin/ ${REVIEWDOG_VERSION}
+RUN apk add --no-cache bash
+
+RUN wget -O - -q https://raw.githubusercontent.com/reviewdog/nightly/master/install.sh | sh -s -- -b /usr/local/bin/ ${REVIEWDOG_VERSION}
```
Fixed in d3c28ac.

rickstaa commented on entrypoint.sh:
```diff
@@ -1,19 +1,21 @@
-remark --frail --quiet --use=remark-preset-lint-recommended . ${INPUT_REMARK_FLAGS} 2>&1 |
+# NOTE: ${VAR,,} Is bash 4.0 syntax to make strings lowercase.
+echo "[action-remark-lint] Checking markdown code with the remark-lint linter and reviewdog..."
+remark --use=remark-preset-lint-recommended . ${INPUT_REMARK_ARGS} 2>&1 |
```
Done by design.
